flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java

```diff
@@ -705,6 +680,47 @@ public int run(String[] args) throws CliArgsException, FlinkException {
 		return 0;
 	}
 
+	private void shutdownCluster(AbstractYarnClusterDescriptor yarnClusterDescriptor,
+		ClusterClient clusterClient, ScheduledExecutorService scheduledExecutorService,
+		YarnApplicationStatusMonitor yarnApplicationStatusMonitor, ApplicationId yarnApplicationId) {
+		try {
+			yarnApplicationStatusMonitor.close();
+		} catch (Exception e) {
+			LOG.info("Could not properly close the Yarn application status monitor.", e);
+		}
+
+		clusterClient.shutDownCluster();
+
+		try {
+			clusterClient.shutdown();
+		} catch (Exception e) {
+			LOG.info("Could not properly shutdown cluster client.", e);
+		}
+
+		// shut down the scheduled executor service
+		ExecutorUtils.gracefulShutdown(
+			1000L,
+			TimeUnit.MILLISECONDS,
+			scheduledExecutorService);
+
+		deleteYarnPropertiesFile();
+
+		ApplicationReport applicationReport;
+
+		try {
+			applicationReport = yarnClusterDescriptor
+				.getYarnClient()
+				.getApplicationReport(yarnApplicationId);
+		} catch (YarnException | IOException e) {
+			LOG.info("Could not log the final application report.", e);
+			applicationReport = null;
+		}
+
+		if (applicationReport != null) {
+			logFinalApplicationReport(applicationReport);
+		}
```

Comment:
I think logging of the final `ApplicationReport` should only happen if we properly close the session. If the shutdown hook is executed, then this does not need to be executed. I guess we could move this part into a separate method and then call it only in the finally block.

Reply:
Make sense to me. Thanks.